core/lib/spree/core/number_generator.rb

```diff
@@ -35,17 +30,13 @@ def included(host)
       def generate_permalink(host)
         length = @length
 
-        loop do
-          candidate = new_candidate(length)
-          return candidate unless host.exists?(number: candidate)
-
-          # If over half of all possible options are taken add another digit.
-          length += 1 if host.count > Rational(BASE**length, 2)
+        candidate = @letters ? 36 : 10
+        number = loop do
+          random_token = SecureRandom.random_number(candidate**length).to_s(candidate).rjust(length, "0")
+          break random_token unless host.exists?(number: random_token)
         end
-      end
 
-      def new_candidate(length)
-        @prefix + length.times.map { @candidates.sample(random: @random) }.join
+        return @prefix + number
```

Comment:
@wuboy0307 last line in method is a result so you don't need to write `return sth`